File: AMF5/src/main/scala/scalaPlatform/CustomResourceLoader.scala
package scalaPlatform

import amf.core.client.common.remote.Content
import amf.core.client.scala.resource.ResourceLoader
import resource._

import java.util.regex.Pattern
import scala.concurrent.Future
import scala.io.Source

class CustomResourceLoader(private val pathPattern: Pattern) extends ResourceLoader {

  /** Fetch specified resource and return associated content. Resource should have been previously accepted. */
  override def fetch(resource: String): Future[Content] = {
    val normalizedPath = resource.substring(pathPattern.pattern.length - 1)
    var content = ""
    for { source <- managed(Source.fromFile(normalizedPath)) } {
      content = source.mkString
    }
    Future.successful(new Content(content, resource))
  }

  /** Accepts specified resource. */
  override def accepts(resource: String): Boolean = {
    if (resource == null || resource.isEmpty) return false
    val matcher = pathPattern.matcher(resource)
    matcher.find
  }
}

File: AMF5/src/main/java/javaPlatform/CustomResourceLoader.java
package javaPlatform;

import amf.core.client.common.remote.Content;
import amf.core.client.platform.resource.FileResourceLoader;
import amf.core.client.platform.resource.ResourceLoader;

import java.io.File;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class CustomResourceLoader implements ResourceLoader {
    private final FileResourceLoader resourceLoader = new FileResourceLoader();
    private final Pattern CUSTOM_PATH_PATTERN = Pattern.compile("^CustomProtocol/");

    @Override
    public CompletableFuture<Content> fetch(String path) {
        final String normalizedPath = path.substring(CUSTOM_PATH_PATTERN.pattern().length() - 1);
        return resourceLoader.fetch(new File(normalizedPath).getAbsolutePath());
    }

    @Override
    public boolean accepts(String resource) {
        if (resource == null || resource.isEmpty()) return false;
        final Matcher matcher = CUSTOM_PATH_PATTERN.matcher(resource);
        return matcher.find();
    }
}
